src/c++/library/http_client.cc

```diff
@@ -1898,47 +1899,58 @@ InferenceServerHttpClient::AsyncTransfer()
       return !this->ongoing_async_requests_.empty();
     });
 
-    curl_multi_perform(multi_handle_, &place_holder);
-    while ((msg = curl_multi_info_read(multi_handle_, &place_holder))) {
-      uintptr_t identifier = reinterpret_cast<uintptr_t>(msg->easy_handle);
-      auto itr = ongoing_async_requests_.find(identifier);
-      // This shouldn't happen
-      if (itr == ongoing_async_requests_.end()) {
-        std::cerr << "Unexpected error: received completed request that is not "
-                     "in the list of asynchronous requests"
-                  << std::endl;
-        curl_multi_remove_handle(multi_handle_, msg->easy_handle);
-        curl_easy_cleanup(msg->easy_handle);
-        continue;
-      }
+    CURLMcode mc = curl_multi_perform(multi_handle_, &place_holder);
+    int numfds;
+    if (mc == CURLM_OK) {
+      // Wait for activity. Timeout should be capped by server_timeout_
+      // of InferOptions
+      mc = curl_multi_poll(multi_handle_, NULL, 0, INT_MAX, &numfds);
```

Comment:
Should we change `INT_MAX` to `timeout_ms` or it obeys the timeout set in L1805?

Reply:
> It will block until activity is detected on at least one of the handles or timeout_ms has passed. Alternatively, if the multi handle has a pending internal timeout that has a shorter expiry time than timeout_ms, that shorter time will be used instead to make sure timeout accuracy is reasonably kept. 

Source: https://curl.se/libcurl/c/curl_multi_poll.html

I think it is still a good idea to change INT_MAX to timeout_ms.

Reply:
I was wondering what would happen if the response was received on client before we begin polling.

Reply:
@Tabrizian and @tanmayv25 re:

> Should we change INT_MAX to timeout_ms or it obeys the timeout set in L1805?

I have thought about this and there's a couple of things:
1. We have not only `InferOptions::client_timeout_` but also `InferOptions::server_timeout_`. I think we should wait for the shorter of the two. 
2. `InferOptions` is set for each `AsyncInfer` call. Given the return time for each of the request could be different (eg: request1 has `client_timeout` of 5m, request2 has `client_timeout` of 1m, what if request 1 returns later, but the wait time is set to 1m, the loop will not wait for long enough then, right?

Reply:
Let's assume that the user hasn't provided timeout with a single request and the response is delivered before  `curl_multi_poll` is initiated. Wouldn't this cause a hang as the timeout would stay INT_MAX?

Does it make sense to put default timeout to be something more reasonable like 1sec?

Reply:
I think reasonableness is with respect to the input request. The `Transfer` loop is always run before the request is even pre-processed. I think this opens up the opportunity for us keeping have to change what "reasonable" means for future runs, if we want to tune Triton's performance.

Reply:
> Let's assume that the user hasn't provided timeout with a single request and the response is delivered before curl_multi_poll is initiated. Wouldn't this cause a hang as the timeout would stay INT_MAX?

Looks like this could be an issue. If there are zero file descriptors, curl_multi_poll can block indefinitely. I think we probably want to use [curl_multi_wait ](https://curl.se/libcurl/c/curl_multi_wait.html) instead:

> If no extra file descriptors are provided and libcurl has no file descriptor to offer to wait for, this function will return immediately. (Try [curl_multi_poll](https://curl.se/libcurl/c/curl_multi_poll.html) instead if you rather avoid this behavior.)

@jbkyang-nvi I think we don't need to have the `shortest_timeout_ms_` interface. The `server_timeout` is taken care of in the server side. We only have a single global timeout which we can set in `curl_multi_wait`. Just make sure that L0_client_timeout passes after these changes.